#18 패스워드 인코딩

0hun opened the pull request.

0hun committed:
#18 패스워드 인코딩

0hun committed:
#18 패스워드 인코딩

PasswordEncoder를 구현한 BCryptPasswordEncoder를 Bean으로 등록

f-lab-dev commented on pom.xml:
```diff
@@ -89,6 +89,24 @@
+        <!--스프링시큐리티 web 라이브러리-->
+        <dependency>
+            <groupId>org.springframework.security</groupId>
+            <artifactId>spring-security-web</artifactId>
```
그냥 비밀번호 암호화하는데 spring security까지 쓸 필요가 있을까요?

0hun replied:
없을것 같습니다. 시큐리티에서 제공하는 기능이라 넣었는데, 필요한 lib만 추가해야되는건가요?

f-lab-dev commented on src/main/java/com/younghun/hibusgo/domain/Account.java:
```diff
@@ -42,4 +43,8 @@
     private LocalDateTime updatedAt;
+
+    public void encodePassword(PasswordEncoder passwordEncoder) {
+        this.password = passwordEncoder.encode(this.password);
```
이 클래스는 데이터를 담기만 하는 역할을 갖는 클래스인데 비밀번호를 암호화하는 책임도 갖게되었네요.

0hun committed:
#18 패스워드 인코딩

시큐리티 필요한 라이브러만 사용하도록 수정
Account클래스가 비밀번호 암호화하는 책임을 갖지않고, 외부에서 암호화한 비밀번호만
내부 함수로 받도록 수정

f-lab-dev commented on src/main/java/com/younghun/hibusgo/domain/Account.java:
```diff
@@ -44,7 +44,7 @@
 
-    public void encodePassword(PasswordEncoder passwordEncoder) {
-        this.password = passwordEncoder.encode(this.password);
+    public void setEncodePassword(String encodePassword) {
```
굳이 `setEncodePassword`라는 메소드를 따로 빼야할까요?
이 클래스의 관심사는 패스워드를 담는 것이지 `패스워드가 암호화되었냐` 는 관심사가 아닌 듯 하네요~
그리고 인자로 받는 패스워드가 암호화되었다는 검증도 할 수 없으니 `setPassword`와 다를 게 없는 것 같습니다.

0hun committed:
#18 패스워드 인코딩

암호화된 패스워드를 담기위해 함수를 따로 빼지않고
빌더 패턴으로 넘기도록 수정

0hun committed:
#18 패스워드 인코딩

Account 클래스에서 사용하지 않는 PasswordEncoder import 제거

0hun committed:
#18 패스워드 인코딩

계정 생성할때 immutable이 깨지지 않도록 암호화된 패스워드를 내부 함수 및 setter를 이용해서 패스워드를 변경하는 것이 아닌 새로운 객체를 생성해서 패스워드를 변경하는 식으로 수정

f-lab-dev commented on src/main/java/com/younghun/hibusgo/domain/Account.java:
```diff
@@ -42,4 +42,15 @@
     // 최종 수정일
     private LocalDateTime updatedAt;
+
+    public Account passwordEncodedAccount(Account account, String encodePassword) {
```
- 변수명은 `encodePassword`지만 암호화되지 않은 값을 넣는 것은 어떻게 방지할 수 있나요?
- 값을 담는 클래스가 암호화되었는지 여부를 신경써야할까요?
- 이 메소드는 copy가 역할인 것 같은데 Account 자신의 정보를 카피하는게 아니라 다른 객체를 받아서 카피합니다. 그럼 static으로 선언하는게 낫지 않을까요?

0hun committed:
#18 패스워드 인코딩

계졍생성 단계에서 패스워드 인코딩시 다른 객체를 넘겨주는것이 아닌 자신의 정보를 카피하도록 수정.
비밀번호를 인코딩하지 않고 넘겨서 인코딩 하도록 수정

0hun committed:
#18 패스워드 인코딩

이전 패스워드를 인코딩하여 나온 값과 먼저 인코딩된 패스워드를 비교하여,
true일 경우 인코딩된 패스워드를 넘겨 회원을 추가하도록 수정

f-lab-dev commented on src/main/java/com/younghun/hibusgo/service/AccountService.java:
```diff
@@ -24,8 +24,8 @@ public Account findById(String id) {
-        String encodePassword = account.getPassword();
-        Account newAccount = account.passwordEncodedAccount(account, encodePassword);
+        String encodePassword = passwordEncoder.encode(account.getPassword());
+        Account newAccount = account.passwordEncodedAccount(encodePassword);
```
아무것도 모르는 사람이 봤을 때 `passwordEncodedAccount`라는 이름으로 **비밀번호만 바뀐 새로운 객체를 복사해서 돌려받는다**라는 것을 유추할 수 있을까요?

f-lab-dev commented on src/main/java/com/younghun/hibusgo/service/AccountService.java:
```diff
@@ -25,8 +25,12 @@ public Account findById(String id) {
+        if (passwordEncoder.matches(rawPassword, encodePassword)) {
+            Account newAccount = account.passwordEncodedAccount(encodePassword);
+            accountMapper.addAccount(newAccount);
+        }
```
만약 매치하지 않으면 에러상황이니 적절한 처리를 해줘야겠습니다.
이 메소드를 호출한 사람은 아무런 메세지가 없으니 실패했더라도 성공한 것으로 인지할 수 있습니다

0hun committed:
Merge branch 'develop' into issue/18

0hun committed:
#18 패스워드 인코딩

패스워드 비교 실패시 회원 가입을 실패한것이므로 에러 처리

0hun committed:
Merge branch 'issue/18' of https://github.com/f-lab-edu/hi-bus-go into issue/18

0hun committed:
#18 패스워드 인코딩

비밀번호만 인코딩하여 객체를 복사한다는 의미로 함수 이름 변경

0hun committed:
#18 패스워드 인코딩

회원 찾기, 회원 수정시 비밀번호 인코딩 하도록 수정

f-lab-dev commented on src/main/java/com/younghun/hibusgo/service/AccountService.java:
```diff
@@ -27,10 +29,12 @@ public void addAccount(Account account) {
-            Account newAccount = account.passwordEncodedAccount(encodePassword);
-            accountMapper.addAccount(newAccount);
+        if (!passwordEncoder.matches(rawPassword, encodePassword)) {
+            throw new HttpServerErrorException(HttpStatus.INTERNAL_SERVER_ERROR, "암호화된 비밀번호가 일치하지 않아, 회원 가입에 실패하였습니다.");
```
비즈니스로직 레이어인 서비스에서 HTTP에 대해 알아야할까요?

f-lab-dev replied:
아직 바뀌지 않은 것 같네요~
HTTP에 대해 service 레이어가 알아야할까요? 다른 예외 타입을 사용해주세요~ 기왕이면 자바 표준예외면 좋겠습니다

0hun replied:
넵 수정하겠습니다.

0hun committed:
#18 패스워드 인코딩

비지니스 레이어에서 http에 대해 알필요없으므로, 컨트롤러에서 에러처리

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/AccountController.java:
```diff
@@ -79,7 +80,11 @@ public void passwordInitBinder(WebDataBinder webDataBinder) {
+        boolean isAddedAccount = accountService.addAccount(account);
+
+        if (!isAddedAccount) {
+            throw new HttpServerErrorException(HttpStatus.INTERNAL_SERVER_ERROR, "암호화된 비밀번호가 일치하지 않아, 회원 가입에 실패하였습니다.");
```
여기선 굳이 예외를 던질 필요 없이 `INTERNAL_SERVER_ERROR` 엔티티를 리턴해주면 되지 않을까요?

f-lab-dev commented on src/main/java/com/younghun/hibusgo/service/AccountService.java:
```diff
@@ -25,16 +23,17 @@ public Account findById(long id) {
 
         if (!passwordEncoder.matches(rawPassword, encodePassword)) {
-            throw new HttpServerErrorException(HttpStatus.INTERNAL_SERVER_ERROR, "암호화된 비밀번호가 일치하지 않아, 회원 가입에 실패하였습니다.");
+            return false;
```
변환된 비밀번호가 맞지 않으면 이거는 우리가 정의한 흐름에 따르면 정상적인 상황일까요? 혹은 예외상황일까요?

0hun replied:
예외상황입니다.

0hun replied:
예외 상황입니다.

f-lab-dev replied:
네 그러면 여기는 서비스레이어이므로 예외 던지는거로 다시 바꿔주세요~

0hun committed:
#18 리뷰 반영

유저 추가하는데 실패했을경우 예외를 던지는 것이 아니리,
미리 정의한 500에러를 담고있는 ResponseEntity 객체를 리턴하도록 수정

0hun committed:
#18 리뷰 반영

비밀번호가 일치하지 않을시 예외상황이므로 서비스 레이어에서 예외를 던지도록 수정

0hun committed:
#18 리뷰 반영

서비스 레이어에서는 HTTP에 대해 알필요가 없으므로, 자바 표준 예외를 던지도록 수정

0hun committed:
#18 리뷰 반영

회원 추가시 암호화 변경시 암호화 체크하지 않도록 수정
로그인시 아이디 비밀번호 체크하도록 수정

0hun committed:
#18 쿼리 수정

쿼리 깔끔하게 보이도록 수정

f-lab-dev commented on src/main/java/com/younghun/hibusgo/validator/LoginDtoValidator.java:
```diff
@@ -0,0 +1,56 @@
+      errors.rejectValue("Account", "wrong.value", "가입하지 않은 아이디이거나, 잘못된 비밀번호입니다.");
+    }
+
+    if (!existIngAccount.get().getPassword().equals(encodePassword)) {
```
아이디, 비밀번호에 대한 검증은 비즈니스적인 요구사항입니다. controller에서 직접 호출되는 validator에서 이거를 검증해야할까요?
기존대로 Service에서 진행하는게 좋아보입니다~

0hun committed:
#18 리뷰 반영

아이디 또는 비밀번호가 맞지 않을 경우 메세지를 리턴하도록 수정

0hun committed:
#18 패스워드 인코딩

로그인시 아이디, 비밀번호가 맞지 않을경우 에러를 던지도록 수정

0hun committed:
#18 로그인 체크 메소드 수정

로그인 계정을 가져오는 메소드가 중복되어 수정
아이디, 비밀번호가 맞지 않을경우 어차피 마지막에 계정을 가지오지 못하므로,
계정을 가져오는 메소드를 하나만 사용하도록 수정
계정이 존재 하지 않을경우 잘못된 요청을 의미하는 400 code return
에러를 던질경우 ResponseEntity의 헤더값에 들어가서 메세지가 들어가지 않아
500에러라는 것만 알수있음(메세지가 전달되지 않음)
(페이스북 및 카카오는 400 code return)
